### PR TITLE
Refactor route protection 

### PR DESCRIPTION
**Pull request by Rodgers-M**
**What does this PR do?**

- improves the app request pre_hook and adds app url prefix

**Description of tasks to be completed**

- ensure the pre_request hook returns appropriate responses for unauthenticated requests
- remove duplicate checking of user object existnce of user object in each route 
- define a midlwarwre to add  a url prefix to the app

**Relevant tracker stories**

- 154453559

### CODE REVIEW

File: app/events/views.py [hunk 1]
```diff
@@ -30,169 +30,155 @@ def validate_date(event_date):
...
-			if not existing_event:
-				#create the event if does not exist
-				event = Events(name=name, description=description, category=category, \
-					location=location, event_date=event_date, created_by=created_by)
-				event.save()
-				res = event.to_json()
-				return jsonify(res), 201
-			return jsonify({"message" : "you have a similar event in the same location"}), 302
-	return jsonify({"message" : "please login or register  to create an event"}), 401
+	event_details = request.get_json()
```
Comment (frankip) [inline app/events/views.py:33]: why did you remove the user verification before creating an event, A user is supposed to log in first before he can be able to create an event, or where are you doing that validation?

File: app/auth/views.py [hunk 2]
```diff
@@ -19,6 +19,10 @@ def before_request():
 					#find the user with the id on the token
 					user = User.query.filter_by(id=res).first()
 					g.user = user
+					return
```
Comment (frankip) [inline app/auth/views.py:22]: an empty return statement

File: app/middleware.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+"""This module defines a classs with a call method for adding route url"""
+class PrefixMiddleware(object):
```
Comment (frankip) [inline app/middleware.py:2]: great work testing the prefix route

**Review by frankip — COMMENTED** [review]
I don't get why you removed all the user validation, even after creating the method, there should be a way to check if the g.user is valid before proceeding to fo the normal CRUD functionalities

File: app/auth/views.py (hunk 2 above)
Comment (Rodgers-M) [reply app/auth/views.py:22]: yes, in python return statements means the end of the current process if they return a value. However, if it returns nothing or `None` the process does not stop. 

in this scenario, this is a pre_request hook, this function runs before every request is processed to ensure a user is authenticated. if a user is authenticated, it returns nothing, meaning the app should go ahead and process the request.

File: app/events/views.py (hunk 1 above)
Comment (Rodgers-M) [reply app/events/views.py:33]: the validation is done in  `@auth.before_app_request` in  `auth/views.py. module where the `access_token` is verified and the apropriate response returned.

File: app/events/views.py [hunk 4]
```diff
@@ -30,169 +30,155 @@ def validate_date(event_date):
...
-		result = Events.query.filter(cast(Events.event_date, Date) >=  date.today()).paginate(page, per_page=15, error_out=False)
-		if result.items:
-			event_list = []
-			for event in result.items:
-				evnt = event.to_json()
-				event_list.append(evnt)
-			return jsonify(event_list), 200
-		return jsonify({"message" : "this page has no events, or no events available"}), 200
-	return jsonify({"message" : "please login or register view events"}), 401
+	#fetch the first 15 events based on event date
```
Comment (meshnesh) [inline app/events/views.py:64]: good implementation

**Comment by meshnesh** [comment]
Good implementation reducing code so that you do not repeat your self.

**Review by betsybeth — COMMENTED** [review]
Good work in validating your arguements
